Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.Firestore.DistributedCache/FirestoreCache.cs

```diff
@@ -0,0 +1,323 @@
+                    {
+                        var slidingExpiration =
+                            doc.LastRefresh.GetValueOrDefault()
+                            + TimeSpan.FromSeconds(doc.SlidingExpirationSeconds.Value);
+                        if (slidingExpiration < now)
+                        {
+                            writeBatch.Delete(docSnapshot.Reference,
+                                Precondition.LastUpdated(
+                                    docSnapshot.UpdateTime.GetValueOrDefault()));
+                            batchSize += 1;
+                        }
+                    }
+                }
+                if (batchSize > 0)
+                {
+                    _logger.LogDebug("Collecting {0} cache entries.", batchSize);
+                    await writeBatch.CommitAsync(token);
+                }
+                if (token.IsCancellationRequested)
+                {
+                    return;
+                }
+            } while (batchSize > 0);
+            _logger.LogTrace("End garbage collection.");
+        }
+
+        internal static string GetProjectId()
+        {
+            // Use the service account credentials, if present.
+            GoogleCredential googleCredential = Google.Apis.Auth.OAuth2
+                .GoogleCredential.GetApplicationDefault();
+            if (googleCredential != null)
+            {
+                ICredential credential = googleCredential.UnderlyingCredential;
+                ServiceAccountCredential serviceAccountCredential =
+                    credential as ServiceAccountCredential;
+                if (serviceAccountCredential != null)
+                {
+                    return serviceAccountCredential.ProjectId;
+                }
+            }
+            try
+            {
+                // Query the metadata server.
+                HttpClient http = new HttpClient();
+                http.DefaultRequestHeaders.Add("Metadata-Flavor", "Google");
+                http.BaseAddress = new Uri(
+                    @"http://metadata.google.internal/computeMetadata/v1/project/");
+                return http.GetStringAsync("project-id").Result;
+            }
+            catch (AggregateException e)
+            when (e.InnerException is HttpRequestException)
+            {
+                throw new Exception("Could not find Google project id.  " +
+                    "Run this application in Google Cloud or follow these " +
+                    "instructions to run locally: " +
+                    "https://cloud.google.com/docs/authentication/getting-started",
+                    e.InnerException);
+            }
+        }
+    }
+
+    /// <summary>
+    /// The document object stored in Firestore for each cache value.
+    /// </summary>
+    [FirestoreData]
+    internal class CacheDoc
+    {
+        [FirestoreProperty]
+        public byte[] Value { get; set; }
+
+        //

```

Comment:
Please rewrite as XML documentation (and below) - and ideally provide documentation for LastRefresh as well.

Reply:
Done.